### PR TITLE
regression 4003: test TEE_CopyOperation()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Adds test for TEE_CopyOperation() in the 4003 in the same way as we have
for 4001 and 4002.

Signed-off-by: Tony He <tony.he@armchina.com>
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

Depends on https://github.com/OP-TEE/optee_os/pull/4268 for tests to pass.

### CODE REVIEW

**Comment by jforissier** [comment]
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

File: host/xtest/regression_4000.c [hunk 1]
```diff
@@ -2211,10 +2223,31 @@ static void xtest_tee_test_4003(ADBG_Case_t *c)
...
+				out + out_offs2,
+				&out_size)))
+			goto out;
+
+		out_offs2 += out_size;
+
+		(void)ADBG_EXPECT_BUFFER(c, ciph_cases[n].out,
+					 ciph_cases[n].out_len, out, out_offs2);
+
+
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2242]: remove 1 empty line

File: host/xtest/regression_4000.c [hunk 2]
```diff
@@ -2211,10 +2223,31 @@ static void xtest_tee_test_4003(ADBG_Case_t *c)
...
+			ta_crypt_cmd_cipher_do_final(c, &session, op2,
+				ciph_cases[n].in + ciph_cases[n].in_incr,
+				ciph_cases[n].in_len - ciph_cases[n].in_incr,
+				out + out_offs2,
+				&out_size)))
+			goto out;
+
+		out_offs2 += out_size;
+
+		(void)ADBG_EXPECT_BUFFER(c, ciph_cases[n].out,
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2239]: maybe `(void)` not needed.

File: host/xtest/regression_4000.c [hunk 3]
```diff
@@ -2070,11 +2070,12 @@ static void xtest_tee_test_4003(ADBG_Case_t *c)
 {
 	TEEC_Session session = { };
 	TEE_OperationHandle op = TEE_HANDLE_NULL;
+	TEE_OperationHandle op2 = TEE_HANDLE_NULL;
 	TEE_ObjectHandle key1_handle = TEE_HANDLE_NULL;
 	TEE_ObjectHandle key2_handle = TEE_HANDLE_NULL;
 	uint8_t out[2048] = { };
 	size_t out_size = 0;
-	size_t out_offs = 0;
+	size_t out_offs = 0, out_offs2 = 0;
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2078]: prefer over 2 lines

**Review by etienne-lms — COMMENTED** [review]
minor comments. LGTM.

**Comment by jenswi-linaro** [comment]
Comments addressed, tags applied. I took the LGTM as a Reviewed-by.

**Comment by hejunzju** [comment]
@jenswi-linaro Thanks for the efforts in following up all these stuffs!

B.t.w, would you mind to apply the following patch for 4005 to complete the TEE_CopyOperation() test?

There is one more issue regarding to the AE (decryption mode) in the present TEE_CopyOperation() implementation. 
See [TEE_CopyOperation issue fails the TEE_ALG_AES_CTS cipher OP-TEE/optee_os#4263](https://github.com/OP-TEE/optee_os/issues/4263)

```
diff --git a/host/xtest/regression_4000.c b/host/xtest/regression_4000.c
index a061dc1..26889c7 100644
--- a/host/xtest/regression_4000.c
+++ b/host/xtest/regression_4000.c
@@ -2874,11 +2874,13 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
 {
 	TEEC_Session session = { };
 	TEE_OperationHandle op = TEE_HANDLE_NULL;
+	TEE_OperationHandle op2 = TEE_HANDLE_NULL;
 	TEE_ObjectHandle key_handle = TEE_HANDLE_NULL;
 	TEE_Attribute key_attr = { };
 	uint8_t out[512] = { };
 	size_t out_size = 0;
 	size_t out_offs = 0;
+	size_t out_offs2 = 0;
 	uint32_t ret_orig = 0;
 	size_t n = 0;
 
@@ -2902,6 +2904,12 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
 				key_attr.content.ref.length * 8)))
 			goto out;
 
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			ta_crypt_cmd_allocate_operation(c, &session, &op2,
+				ae_cases[n].algo, ae_cases[n].mode,
+				key_attr.content.ref.length * 8)))
+			goto out;
+
 		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
 			ta_crypt_cmd_allocate_transient_object(c, &session,
 				ae_cases[n].key_type,
@@ -2976,7 +2984,12 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
 			}
 		}
 
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			ta_crypt_cmd_copy_operation(c, &session, op2, op)))
+			goto out;
+
 		out_size = sizeof(out) - out_offs;
+		out_offs2 = out_offs;
 		if (ae_cases[n].mode == TEE_MODE_ENCRYPT) {
 			uint8_t out_tag[64];
 			size_t out_tag_len = MIN(sizeof(out_tag),
@@ -3016,10 +3029,56 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
 				ae_cases[n].ptx_len, out, out_offs);
 		}
 
+		/* test on the copied op2 */
+		out_size = sizeof(out) - out_offs2;
+		memset(out + out_offs2, 0, out_size);
+		if (ae_cases[n].mode == TEE_MODE_ENCRYPT) {
+			uint8_t out_tag[64];
+			size_t out_tag_len = MIN(sizeof(out_tag),
+						 ae_cases[n].tag_len);
+
+			if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+				ta_crypt_cmd_ae_encrypt_final(c, &session, op2,
+					ae_cases[n].ptx + ae_cases[n].in_incr,
+					ae_cases[n].ptx_len -
+						ae_cases[n].in_incr,
+					out + out_offs2,
+					&out_size, out_tag, &out_tag_len)))
+				goto out;
+
+			(void)ADBG_EXPECT_BUFFER(c,
+				ae_cases[n].tag, ae_cases[n].tag_len, out_tag,
+				out_tag_len);
+
+			out_offs2 += out_size;
+
+			(void)ADBG_EXPECT_BUFFER(c, ae_cases[n].ctx,
+				ae_cases[n].ctx_len, out, out_offs2);
+		} else {
+			if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+				ta_crypt_cmd_ae_decrypt_final(c, &session, op2,
+					ae_cases[n].ctx + ae_cases[n].in_incr,
+					ae_cases[n].ctx_len -
+						ae_cases[n].in_incr,
+					out + out_offs2,
+					&out_size, ae_cases[n].tag,
+					ae_cases[n].tag_len)))
+				goto out;
+
+			out_offs2 += out_size;
+
+			(void)ADBG_EXPECT_BUFFER(c, ae_cases[n].ptx,
+				ae_cases[n].ptx_len, out, out_offs2);
+		}
+
 		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
 			ta_crypt_cmd_free_operation(c, &session, op)))
 			goto out;
 
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			ta_crypt_cmd_free_operation(c, &session, op2)))
+			goto out;
+
 		Do_ADBG_EndSubCase(c, NULL);
 	}
 out:
```

**Comment by jenswi-linaro** [comment]
Added one more patch to cover 4005 too.

File: host/xtest/regression_4000.c [hunk 4]
```diff
@@ -2526,10 +2539,56 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
 				ae_cases[n].ptx_len, out, out_offs);
 		}
 
+		/* test on the copied op2 */
+		out_size = sizeof(out) - out_offs2;
+		memset(out + out_offs2, 0, out_size);
+		if (ae_cases[n].mode == TEE_MODE_ENCRYPT) {
+			uint8_t out_tag[64];
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2546]: `= { 0 };

File: host/xtest/regression_4000.c [hunk 5]
```diff
@@ -2526,10 +2539,56 @@ static void xtest_tee_test_4005(ADBG_Case_t *c)
...
+			if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+				ta_crypt_cmd_ae_encrypt_final(c, &session, op2,
+					ae_cases[n].ptx + ae_cases[n].in_incr,
+					ae_cases[n].ptx_len -
+						ae_cases[n].in_incr,
+					out + out_offs2,
+					&out_size, out_tag, &out_tag_len)))
+				goto out;
+
+			(void)ADBG_EXPECT_BUFFER(c,
```
Comment (etienne-lms) [inline host/xtest/regression_4000.c:2559]: remove `(void)`.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with 2 minor comments.

**Comment by jforissier** [comment]
For "regression 4005: test TEE_CopyOperation()":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Comments addresses and tags applied to "regression 4005: test TEE_CopyOperation()"

**Pull request merged**